### PR TITLE
Rework metric registration and tracking

### PR DESCRIPTION
**Pull request by claytono**
This reworks/rewrites the way that metric registration and tracking is handled across all of `statsd_exporter`.  The goal here is to reduce memory and cpu usage, but also to reduce complexity by unifying metric registration with the ttl tracking for expiration.

Some high level notes:

 * Previously metric names and labels were being hashed three times for every event accepted: in the container code, the save label set code and again in the prometheus client libraries.  This unifies the first two and caches the results of `GetMetricWith` to avoid the third.

* This optimizes the label hashing to reduce cpu overhead and memory allocations.  The label hashing code previously showed up high on all profiling done for both CPU and memory allocations

Using the `BenchmarkExporterListener` benchmark, the improvement looks like this.

 Before:
```
cpu: 11,341,797 ns/op
memory allocated: 1,731,119 B/op
memory allocations: 58,028 allocs/op
```

After:
```
cpu: 7,727,517 ns/op
memory allocated: 1,226,725 B/op
memory allocations: 46,026 allocs/op
```

This also adds a benchmark for testing the performance of the `ExporterListener`.

### CODE REVIEW

File: exporter_test.go [hunk 1]
```diff
@@ -724,13 +724,11 @@ func getFloat64(metrics []*dto.MetricFamily, name string, labels prometheus.Labe
 	}
 
 	var metric *dto.Metric
-	sortedLabelNames := getSortedLabelNames(labels)
-	labelsHash := hashNameAndLabels(name, sortedLabelNames, labels)
```
Comment (claytono) [inline exporter_test.go]: `getSortedLabelNames` and `hashNameAndLabels` no longer exist.  Printing the labels into a string is slow but produces a sorted representation with less complexity.

**Comment by claytono** [comment]
I'm going to do some more testing with this, but this is a big change, so I wanted to go ahead and put it up for review.  So far my testing looks pretty good.

**Comment by claytono** [comment]
I found an error with this during a test deploy that I'm going to investigate.

**Comment by claytono** [comment]
I fixed a few issues that I found in my testing environment.  It's looking good now.  I'll check back on it after the weekend and see what things look like.

File: registry.go [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+				metric.vectors[rm.vecKey].refCount--
+				delete(metric.metrics, hash)
+			}
+		}
+	}
+}
+
+// Calculates a hash of both the label names and the label names and values.
+func (r *registry) hashLabels(labels prometheus.Labels) (labelHash, []string) {
+	r.hasher.Reset()
```
[the remark this answers is not in the document]
Comment (claytono) [reply registry.go]: No, I don't think we do.  The use here is functionally the same as the global hash instances in the old code.

File: exporter_test.go [hunk 3]
```diff
@@ -536,6 +536,51 @@ func TestHistogramUnits(t *testing.T) {
...
+			labels:     labels,
+		},
+		&CounterEvent{
+			metricName: name,
+			value:      1,
+			labels:     labels,
+		},
+	}
+	events <- c
+	events <- Events{}
```
[the remark this answers is not in the document]
Comment (claytono) [reply exporter_test.go]: It's not.  I had copied another test when starting this one, and I accidentally left this in place.  I just pushed a new commit to remove the second one.

[the remark this answers is not in the document]
Comment (claytono) [reply exporter_test.go]: Actually, it looks like the test fails without it, but I'm not sure why.  I'll investigate.

[the remark this answers is not in the document]
Comment (claytono) [reply exporter_test.go]: Ah, I see why.  The channel isn't buffered.  by pushing an empty event into the channel, it ensures the exporter has consumed the first one before validating the metric.  I'll add a comment explaining it.

**Comment by matthiasr** [comment]
thank you!

**Pull request merged**